library/core/src/escape.rs

```diff
@@ -149,77 +151,197 @@ const fn escape_unicode<const N: usize>(c: char) -> ([ascii::Char; N], Range<u8>
+#[derive(Clone, Copy)]
+#[non_exhaustive]
+pub(crate) struct AlwaysEscaped;
+
+/// Marker type to indicate that the character may be escaped,
+/// use to optimized the iterator implementation.
+#[derive(Clone, Copy)]
+#[non_exhaustive]
+pub(crate) struct MaybeEscaped;
+
 /// An iterator over an fixed-size array.
 ///
-/// This is essentially equivalent to array’s IntoIter except that indexes are
-/// limited to u8 to reduce size of the structure.
-#[derive(Clone, Debug)]
-pub(crate) struct EscapeIterInner<const N: usize> {
+/// This is essentially equivalent to array’s `IntoIter` except that
+/// indices are stored as `u8` to reduce size of the structure.
+#[derive(Clone)]
+pub(crate) struct EscapeIterInner<const N: usize, ESCAPING> {
     // The element type ensures this is always ASCII, and thus also valid UTF-8.
-    data: [ascii::Char; N],
+    // Invariant: `N < 128`. For non-printable characters, contains the escaped
+    // representation as ASCII characters. For printable characters, contains the
+    // character itself.
+    data: MaybeEscapedCharacter<N>,
 
-    // Invariant: `alive.start <= alive.end <= N`
+    // Invariant: For non-printable characters, `alive.start <= alive.end <= N`,
+    // for printable characters, `alive.end >= alive.start >= 128`.
     alive: Range<u8>,
+
+    escaping: PhantomData<ESCAPING>,
 }
 
-impl<const N: usize> EscapeIterInner<N> {
+impl<const N: usize, ESCAPING> EscapeIterInner<N, ESCAPING> {
+    #[inline]
+    const fn new(data: MaybeEscapedCharacter<N>, alive: Range<u8>) -> Self {
+        const { assert!(N < 128) };
+        Self { data, alive, escaping: PhantomData }
+    }
+
     pub(crate) const fn backslash(c: ascii::Char) -> Self {
         let (data, range) = backslash(c);
-        Self { data, alive: range }
+        Self::new(MaybeEscapedCharacter { escaped: data }, range)
     }
 
     pub(crate) const fn ascii(c: u8) -> Self {
         let (data, range) = escape_ascii(c);
-        Self { data, alive: range }
+        Self::new(MaybeEscapedCharacter { escaped: data }, range)
     }
 
     pub(crate) const fn unicode(c: char) -> Self {
         let (data, range) = escape_unicode(c);
-        Self { data, alive: range }
+        Self::new(MaybeEscapedCharacter { escaped: data }, range)
     }
 
     #[inline]
     pub(crate) const fn empty() -> Self {
-        Self { data: [ascii::Char::Null; N], alive: 0..0 }
+        Self::new(MaybeEscapedCharacter { escaped: [ascii::Char::Null; N] }, 0..0)
     }
 
+    /// # Safety
+    ///
+    /// The caller must ensure that `self` contains an escape sequence.
     #[inline]
-    pub(crate) fn as_ascii(&self) -> &[ascii::Char] {
-        // SAFETY: `self.alive` is guaranteed to be a valid range for indexing `self.data`.
+    unsafe fn as_ascii(&self) -> &[ascii::Char] {
+        // SAFETY: `self.data.escaped` contains an escape sequence, as is guaranteed
+        // by the caller, and `self.alive` is guaranteed to be a valid range for
```

Comment:
It seems to me that the top level safety comment should state the more detailed requirement that callers are expected to uphold, and then the inner unsafe use can say that it relies on that.

Tangentially, I'm wondering if `escaped` would not be better as `escape_seq` or `escape`.

Reply:
The second part about `self.alive` is guaranteed by the invariant of the type itself, so the existence of `self` is enough to satisfy it.